snuba/utils/streams/streaming.py

```diff
@@ -0,0 +1,186 @@
+import logging
+import time
+from dataclasses import dataclass
+from typing import Callable, Generic, Mapping, MutableMapping, Optional, TypeVar
+
+from snuba.utils.streams.processing import ProcessingStrategy
+from snuba.utils.streams.types import Message, Partition, TPayload
+
+ProcessingStep = ProcessingStrategy
+
+
+logger = logging.getLogger(__name__)
+
+
+class FilterStep(ProcessingStep[TPayload]):
+    """
+    Determines if a message should be submitted to the next processing step.
+    """
+
+    def __init__(
+        self,
+        function: Callable[[Message[TPayload]], bool],
+        next_step: ProcessingStep[TPayload],
+    ):
+        self.__test_function = function
+        self.__next_step = next_step
+
+    def poll(self) -> None:
+        self.__next_step.poll()
+
+    def submit(self, message: Message[TPayload]) -> None:
+        if self.__test_function(message):
+            self.__next_step.submit(message)
+
+    def close(self) -> None:
+        pass
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__next_step.close()
+        self.__next_step.join(timeout)
+
+
+TTransformed = TypeVar("TTransformed")
+
+
+class TransformStep(ProcessingStep[TPayload]):
```

Comment:
That is my understanding, yes. The relationship between the transformation function and the next step is enforced when type checking the constructor parameters, but the class itself doesn't need to have that type variable bound since since it never accepts it as a parameter or returns it as a return value (like you mentioned.)